notebooks/utils.py

```diff
@@ -0,0 +1,127 @@
+import torch
+from torch import nn
+from torch import optim
+from matplotlib import pyplot as plt
+import numpy as np
+import seaborn as sns
+
+class Swarm_Simulator:
+    def __init__(self):
+        
+        self.simple_funcs = {
+            'exponential':torch.exp,
+            'sinusoidal':torch.sin
+        }
+        
+        self.debug = False
+        
+        self.params = {
+            'domain_min': -1,
+            'domain_max': 3,
+            'num_points': 81,
+            'function': 'exponential',
+            'hidden': 1,
+            'width': 5,
+            'nepoch': 100,
+            'lr': 0.002,
+            'momentum': 0.9,
+            'swarm_size': 20
+            
+        }
+        
+        self.simulations = []
+        
+        
+        self.bees = [{'loss_list':[], 'pred_list': []} for i in range(self.params['swarm_size'])]
+        
+        self.xt = torch.linspace(self.params['domain_min'], self.params['domain_max'], self.params['num_points']).unsqueeze(-1)
+        self.chosen_func = self.simple_funcs[self.params['function']]
+        self.yt = self.chosen_func(self.xt)
+        self.xd = self.xt.detach()
+        self.yd = self.yt.detach()
+        self.get_pred = self.make_predictor(self.xt, self.yt)
+        
+        
+    def reset_swarm(self):
+        self.bees = [{'loss_list':[], 'pred_list': []} for i in range(self.params['swarm_size'])]
+        
+    def rebuild_predictor(self):
+        self.xt = torch.linspace(self.params['domain_min'], self.params['domain_max'], self.params['num_points']).unsqueeze(-1)
+        self.chosen_func = self.simple_funcs[self.params['function']]
+        self.yt = self.chosen_func(self.xt)
+        self.xd = self.xt.detach()
+        self.yd = self.yt.detach()
+        self.get_pred = self.make_predictor(self.xt, self.yt)
+        self.reset_swarm()
+        
+
+    def make_net(self, hidden_depth, width):
+        assert hidden_depth >= 1
+        yield nn.Linear(1, width)
+        yield nn.ReLU()
+        for i in range(hidden_depth-1):
+            yield nn.Linear(width, width)
+            yield nn.ReLU()
+        yield nn.Linear(width, 1)
+    
+    def make_predictor(self, x, y):
+        xt = x.clone()
+        yt = y.clone()
+        def get_pred(num, hidden = self.params['hidden'], width = self.params['width'], nepoch=self.params['nepoch'], lr=self.params['lr'], momentum=self.params['momentum'], debug=self.debug):
```

Comment:
This design pattern is called a closure. It's a way of having class like features without defining a class. They're very handy in notebooks where I want to attach outer scope to a function. Generally speaking you don't combine class and closures - they come from almost opposite schools of programming thought and they're each school's way of achieving similar things.

Reply:
Yea, I think that makes sense.  I could tell as I wrote this that I wasn't really embracing the class fully, but was working to just get something that ran with minimal alterations.